src/twisted/web/test/test_http2.py

```diff
@@ -1425,6 +1425,45 @@ def test_resetAfterBody(self):
         self.assertNotIn(1, a._streamCleanupCallbacks)
 
 
+    def test_RequestRequiringFactorySiteInConstructor(self):
+        """
+        A custom L{Request} subclass that requires the site and factory in the
+        constructor is able to get them.
+        """
+        d = defer.Deferred()
+
+        class SuperRequest(DummyHTTPHandler):
+            def __init__(self, *args, **kwargs):
+                DummyHTTPHandler.__init__(self, *args, **kwargs)
+                d.callback((self.channel.site, self.channel.factory))
+                
+        factory = FrameFactory()
+        transport = StringTransport()
+        connection = H2Connection()
+        httpFactory = http.HTTPFactory()
+        connection.requestFactory = SuperRequest
+
+        # Create some sentinels to look for.
+        connection.factory = httpFactory
+        connection.site = object()
+
+        def validateFactoryAndSite(args):
+            site, factory = args
+            self.assertIs(site, connection.site)
+            self.assertIs(factory, connection.factory)
+        d.addCallback(validateFactoryAndSite)
+
+        requestBytes = factory.clientConnectionPreface()
+        requestBytes += buildRequestBytes(self.getRequestHeaders, [], factory)
+        connection.makeConnection(transport)
+        connection.dataReceived(requestBytes)
```

Comment:
can we have this extracted into something like

```
def connectAndReceive(self, connection):
    factory = FrameFactory()
    transport = StringTransport()
    requestBytes = factory.clientConnectionPreface()
    requestBytes += buildRequestBytes(self.getRequestHeaders, [], factory)
    connection.makeConnection(transport)
    connection.dataReceived(requestBytes)
```

I had to read the code more than once to see what exactly is `factory` what is `httpFactory` and what is ` site, factory = args`

with this extraction, the pieces which are not relevant to the test should be moved away and should make the test easier to read